crates/jarscanner/src/lib.rs

```diff
@@ -25,40 +26,71 @@ fn ends_in_class(file_name: &str) -> bool {
     file_name.ends_with(".class")
 }
 
-fn file_name_to_class_names(file_name: &str) -> Result<Vec<String>, FileNameError> {
-    if non_anon(file_name) && not_in_meta(file_name) && ends_in_class(file_name) {
-        let class_suffix = file_name.strip_suffix(".class").ok_or_else(|| {
-            FileNameError::new(format!("Failed to strip .class suffix for {}", file_name))
-        })?;
-        let final_suffix = class_suffix.strip_suffix("$").unwrap_or(class_suffix);
-
-        let dotted = final_suffix
-            .replace(r"/$", r"/")
-            .replace("$", ".")
-            .replace(r"/", ".");
-        let replace_pkg = dotted.replace(".package", "");
-        if dotted.contains(".package") {
-            Ok(vec![dotted, replace_pkg])
+fn file_name_to_class_names(
+    file_name_ref: &str,
+    buffer: &mut Vec<String>,
+) -> Result<(), FileNameError> {
+    if non_anon(file_name_ref) && not_in_meta(file_name_ref) && ends_in_class(file_name_ref) {
+        let length_of_name = file_name_ref.len();
+        let mut file_name_res = String::with_capacity(length_of_name);
+        let mut was_slash = false;
+        let mut idx = 0;
+        // We know this ends in .class, so we're going to need to track the index of the last 6 chars
+        let end_idx = length_of_name - 6;
+        for file_char in file_name_ref.chars() {
+            if idx == end_idx {
+                break;
+            }
+
+            match file_char {
+                '/' => {
+                    was_slash = true;
+                    file_name_res.push('.')
+                }
+                '$' => {
+                    if idx == end_idx - 1 {
+                        // Do nothing
+                    } else if !was_slash {
+                        was_slash = false;
+                        file_name_res.push('.')
+                    }
+                }
+                _ => {
+                    was_slash = false;
+                    file_name_res.push(file_char)
+                }
+            }
+            idx += 1;
+        }
+
+        if file_name_res.contains(".package") {
+            buffer.push(file_name_res.clone());
+            buffer.push(file_name_res.replace(".package", ""));
+            Ok(())
         } else {
-            Ok(vec![dotted])
+            buffer.push(file_name_res);
+            Ok(())
         }
     } else {
-        Ok(vec![])
+        Ok(())
     }
 }
 
-fn read_zip_archive(input_jar: &PathBuf) -> Result<HashSet<String>, Box<dyn Error>> {
+fn read_zip_archive(input_jar: &PathBuf) -> Result<HashSet<String>, JarscannerError> {
     let file = File::open(input_jar)?;
-    let archive = ZipArchive::new(file)?;
+    let reader = BufReader::with_capacity(32000, file);
+    let archive = ZipArchive::new(reader)?;
 
     let mut result = HashSet::new();
+    let mut buf = Vec::new();
```

Comment:
why do we need to add to a Vec then add those to a HashSet? Can't we just pass the `&mut` of the `HashSet` in instead?